### PR TITLE
[FIX] Added warning for "--ring_size" with 0 as an argument

### LINKED ISSUE #2335: Passing "--ring_size" with 0 as an argument, stucks the execution
#### Describe the bug
Passing `--ring_size` with 0 as an argument, stucks the execution

#### To Reproduce
Run command:
`vw -d train_adf.txt --cb_explore_adf --ring_size 0`

#### Expected behavior
`--ring_size` shouldn't take 0 as an argument.
I think we need to switch `--ring_size` to its default value, and a relevent warning should be displayed, whenever 0 is passed as an argument.

#### Observed Behavior
execution got stuck for an undefined time

#### Environment
What version of VW did you use?
8.8.0
What OS or language did you use?
Linux with python

### PR DESCRIPTION
**Pull request by adityauser**
Fixes #2335.

Switched  `--ring_size` to its default value, and a relevent warning is displayed, whenever 0 is passed as an argument.

### CODE REVIEW

**Comment by mritunjaysharma394** [comment]
Hi, I suggest you to use `git fetch upstream` and then pull it in your forked branch so that your branch is in sync with the base branch. @jackgerrits your thoughts on it?

**Comment by jackgerrits** [comment]
Checking for 0 makes sense. I had a check and if you pass -1, VW crashes. Would you be able to include handling that issue in this PR?

**Comment by adityauser** [comment]
Actually datatype of ring_size is size_t [(see this line](https:github.com/VowpalWabbit/vowpal_wabbit/blob/0eb1ffdd21525a3003e526106915549f95f0a5f4/vowpalwabbit/parse_args.cc#L1327)). So, passing -1 does overflow, on my machine passing 100000 crashes VW.
What should be the upper bound of ring_size? 
Is it related to issue #2123?

**Comment by jackgerrits** [comment]
It is unrelated to #2123. Allowing the input to be negative and them implicitly overflowing is surprising behavior. I think this is the issue that should be fixed. It is my opinion that there should not be a ceiling on the number.

**Comment by adityauser** [comment]
@jackgerrits passing 0.1 throws error: `the argument ('0.1') for option '--ring_size' is invalid`.
It would be better if for -1 also it should throw the same error, but I couldn't make it.
Can you please guide me? I'm not much fimilar with boost options
The solution I could come up with is:
   ```
    bool strict_parse = false;
    int ring_size_int;
    size_t ring_size;
    option_group_definition vw_args("VW options");
    vw_args.add(make_option("ring_size", ring_size_int).default_value(256).help("size of example ring"))
        .add(make_option("strict_parse", strict_parse).help("throw on malformed examples"));
    options.add_and_parse(vw_args);
    if (ring_size_int<0)
    {
        THROW("ring_size should be positive");
    }
    else
    {
        ring_size = ring_size_int;
    }
    if (ring_size == 0)
    {
      all.trace_message << "warning: --ring_size: ring_size can't be 0, switching to the default value of 256"
                          << endl;
      ring_size = 256;
    }
```
In the above code if I change datatype of `ring_size_int` to `long int`, then passing any value of ring_size will thow [this](https://github.com/VowpalWabbit/vowpal_wabbit/blob/a428d8f7954e29e5a5fba76517f47895c2c7528b/vowpalwabbit/options_boost_po.cc#L185) error. It seems to be a bug.

**Comment by jackgerrits** [comment]
The error you showed is because you're passing a non whole number to an option accepting an integer, so boost rightly gives an error. Granted, the error message isn't great. I'm not sure what options boost po gives there.

Your proposed solution is inconsistent. One way throws and the other sets to a default value and prints something. 

To use a type as an option it must be specified here: https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/options_types.h#L33

**Comment by adityauser** [comment]
Why checks are failing?

**Comment by jackgerrits** [comment]
It fails to compile:
> /Users/runner/runners/2.165.2/work/1/s/vowpalwabbit/parse_args.cc:1339:24: error: non-constant-expression cannot be narrowed from type 'int' to 'size_t' (aka 'unsigned long') in initializer list [-Wc++11-narrowing]
    all.p = new parser{ring_size, strict_parse};
                       ^~~~~~~~~

File: vowpalwabbit/parse_args.cc [hunk 1]
```diff
@@ -1325,10 +1325,20 @@ vw& parse_args(options_i& options, trace_message_t trace_listener, void* trace_c
 
     bool strict_parse = false;
     size_t ring_size;
+    int tmp;
```
Comment (jackgerrits) [inline vowpalwabbit/parse_args.cc]: Please use a more descriptive name

File: vowpalwabbit/parse_args.cc [hunk 2]
```diff
@@ -1325,10 +1325,20 @@ vw& parse_args(options_i& options, trace_message_t trace_listener, void* trace_c
...
         .add(make_option("strict_parse", strict_parse).help("throw on malformed examples"));
     options.add_and_parse(vw_args);
+    
+    if (tmp <= 0)
+    {
+      THROW("ring_size should be positive");
+    }
+    else
+    {
+        ring_size = tmp;
```
Comment (jackgerrits) [inline vowpalwabbit/parse_args.cc]: please use a static cast to avoid the warning

File: vowpalwabbit/parse_args.cc [hunk 3]
```diff
@@ -1325,10 +1325,20 @@ vw& parse_args(options_i& options, trace_message_t trace_listener, void* trace_c
...
-    vw_args.add(make_option("ring_size", ring_size).default_value(256).help("size of example ring"))
+    vw_args.add(make_option("ring_size", tmp).default_value(256).help("size of example ring"))
         .add(make_option("strict_parse", strict_parse).help("throw on malformed examples"));
     options.add_and_parse(vw_args);
+    
+    if (tmp <= 0)
+    {
+      THROW("ring_size should be positive");
+    }
+    else
```
Comment (jackgerrits) [inline vowpalwabbit/parse_args.cc]: If you remove the else then you don't need to declare ring_size separate to its initialization.

**Comment by adityauser** [comment]
Is it okay now @jackgerrits ?

**Comment by jackgerrits** [comment]
Looks good, will merge when tests pass.

**Comment by adityauser** [comment]
Negative numbers are also passing flawlessly through parameters (such as: `ftrl_alpha,  ftrl_beta, mem, termination, initial_pass_length, l1, l2, decay_learning_rate arg` ).

**Pull request merged**